both/api/sources/server/stream-chain/stream-types/transform-kobo/rating-rules.ts

```diff
@@ -0,0 +1,164 @@
+import { isMatch, isEqual, entries, get, intersection } from 'lodash';
+
+// TODO align with ac format when available, add operator support
+export interface Quantity {
+  value: number;
+  unit: string;
+}
+
+// rule types
+type Comparable = number | string | Quantity;
+// checks if a property is null or has a value
+type ExistsValue = { $exists: boolean };
+// checks if a property is undefined or has a value (including null)
+type DefinedValue = { $defined: boolean };
+type LessThan = { $lt: Comparable };
+type LessThanEquals = { $lte: Comparable };
+type GreaterThan = { $gt: Comparable };
+type GreaterThanEquals = { $gte: Comparable };
+type Equals = { $eq: Comparable };
+type NotEquals = { $ne: Comparable };
+type MatchValue = string | number | undefined | null | boolean |
+  DefinedValue | ExistsValue |
+  LessThan | LessThanEquals | GreaterThan | GreaterThanEquals | Equals | NotEquals;
+type MatchRule = {
+  [key: string]: MatchValue,
+};
+type OrRule = {
+  $or: ReadonlyArray<Rule>,
+};
+export type Rule = OrRule | MatchRule;
+
+// three valued logic for a11y
+export type RuleEvaluationResult = 'true' | 'false' | 'unknown';
+
+// combine multiple rules with a three valued or, with the order of true > false > unknown
+// this does not align with Kleene and Priest logics, but knowing that a rule does not apply
+// is strong enough reason to determine that the or should be fals
+function evaluateOrRule(data: {}, orRule: OrRule): RuleEvaluationResult  {
+  let finalResult: RuleEvaluationResult = 'unknown';
+
+  for (const rule of orRule.$or) {
+    const result = evaluateRule(data, rule);
+
+    if (result === 'true') {
+      return 'true';
+    }
+    // apply no if found
+    if (finalResult === 'unknown') {
+      finalResult = result;
+    }
+  }
+  return finalResult;
+}
+
+// read the value out of a quantity
+function getQuantityValue(a: string | number | Quantity) : number | string {
+  let aValue: number | string = 0;
+  if (typeof a === 'object') {
+    // todo use better conversion in the future
+    const multiplier = a.unit === 'inch' ? 2.54 : 1;
+    aValue = a.value * multiplier;
+  } else {
+    aValue = a;
+  }
+  return aValue;
+}
+
+// the allowed operator types for comparisons
+type Operators = '$eq' | '$lt' | '$lte' | '$gt' | '$gte' | '$ne';
+const allowedOperators: ReadonlyArray<Operators> =
+  Object.freeze(['$eq', '$lt', '$lte', '$gt', '$gte', '$ne'] as Operators[]);
+
+// compare two values using an operator, if they are quantities, read the underlying value
+function compareByOperator(
+    first: Comparable,
+    second: Comparable,
+    operator: Operators = '$eq'): boolean {
+  const a = getQuantityValue(first);
```

Comment:
-1: This should ensure the units are the same.
If this would be unit-agnostic with js-quantities, it could work directly for any AC object, not only Kobo :)

Reply:
As discussed, let's do this when we move this into ac-format